### PR TITLE
[AsmParser] Add support for reading incomplete IR (part 1)

### PR DESCRIPTION
**Pull request by nikic**
Add an `-allow-incomplete-ir` flag to the IR parser, which allows reading IR with missing declarations. This is intended to produce a best-effort interpretation of the IR, along the same lines of what we would manually do when taking, for example, a function from `-print-after-all` output and fixing it up to be valid IR.

This patch only supports dropping references to undeclared metadata, either by dropping metadata attachments from instructions/functions, or by dropping calls to certain intrinsics (like debug intrinsics). I will implement support for inserting missing function/global declarations in a followup patch.

We don't have real use lists for metadata, so the approach here is to iterate over the whole IR and identify metadata that needs to be dropped. This does not support all possible cases, but should handle anything that's relevant for the function-only IR use case.

### CODE REVIEW

File: llvm/lib/AsmParser/LLParser.cpp [hunk 1]
```diff
@@ -269,10 +330,14 @@ bool LLParser::validateEndOfModule(bool UpgradeDebugInfo) {
 
   for (auto *Inst : InstsWithTBAATag) {
     MDNode *MD = Inst->getMetadata(LLVMContext::MD_tbaa);
-    assert(MD && "UpgradeInstWithTBAATag should have a TBAA tag");
-    auto *UpgradedMD = UpgradeTBAANode(*MD);
-    if (MD != UpgradedMD)
-      Inst->setMetadata(LLVMContext::MD_tbaa, UpgradedMD);
+    // With incomplete IR, the tbaa metadata may have been dropped.
+    if (!AllowIncompleteIR)
+      assert(MD && "UpgradeInstWithTBAATag should have a TBAA tag");
```
Comment (dtcxzyw) [inline llvm/lib/AsmParser/LLParser.cpp:335]: Use `return error(...)` instead?

Comment (nikic) [reply llvm/lib/AsmParser/LLParser.cpp:335]: This is an internal consistency assertion -- it should not be possible to trigger such an error on any input.

**Review by nhaehnle — APPROVED** [review]
Yeah, I can see how this can be convenient sometimes. LGTM.

(I wonder if this can be made more discoverable. Perhaps by outputting a hint about this option when undefined metadata is encountered.)

**Review by fhahn — COMMENTED** [review]
This should be very useful thanks! Might be good to share on discourse to make more people aware of this :)

File: llvm/lib/AsmParser/LLParser.cpp [hunk 2]
```diff
@@ -123,6 +130,57 @@ void LLParser::restoreParsingState(const SlotMapping *Slots) {
         std::make_pair(I.first, std::make_pair(I.second, LocTy())));
 }
 
+void LLParser::dropUnknownMetadataReferences() {
+  auto Pred = [](unsigned MDKind, MDNode *Node) { return Node->isTemporary(); };
+  for (Function &F : *M) {
+    F.eraseMetadataIf(Pred);
+    for (BasicBlock &BB : F) {
```
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/AsmParser/LLParser.cpp]: Done

File: llvm/lib/AsmParser/LLParser.cpp [hunk 3]
```diff
@@ -123,6 +130,57 @@ void LLParser::restoreParsingState(const SlotMapping *Slots) {
...
+      for (Instruction &I : make_early_inc_range(BB)) {
+        I.eraseMetadataIf(Pred);
+
+        if (auto *II = dyn_cast<IntrinsicInst>(&I)) {
```
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/lib/AsmParser/LLParser.cpp]: I've opted to move this logic into a separate function instead, so we can use early return.

**Comment by nikic** [comment]
> This should be very useful thanks! Might be good to share on discourse to make more people aware of this :)

To follow up on this comment, I've shared this on discourse here: https://discourse.llvm.org/t/recent-improvements-to-the-ir-parser/77366?u=nikic

**Pull request merged**